### PR TITLE
Adding new cog - splitify

### PR DESCRIPTION
**Pull request by vincepy**
## Relevant Issues
No issues, first time adding this command.

## Description
Merged splitifer and a discord command for it together, added neat descriptions for each function and the discord command, fixed all the styling errors.

## Reasoning
Implemented everything using 8bitify cog as a base example.

## Screenshots
Command usage:

image
image

Error messages:

## Additional Details
Nothing to add.

## Did you:

- [x] Join the [**Python Discord Community**](https://discord.gg/python)?
- [x] If dependencies have been added or updated, run `pipenv lock`?
- [x] **Lint your code** (`pipenv run lint`)?

### CODE REVIEW

File: bot/exts/evergreen/splitify.py
Comment (ToxicKidz) [inline bot/exts/evergreen/splitify.py:80]: Maybe there should be a default value like `16`?

File: bot/exts/evergreen/splitify.py
Comment (ToxicKidz) [inline bot/exts/evergreen/splitify.py:108]: I think this part should be in a separate function and it should be [ran in the executor](https://docs.python.org/3/library/asyncio-eventloop.html#asyncio.loop.run_in_executor).

File: bot/exts/evergreen/splitify.py
Comment (ToxicKidz) [inline bot/exts/evergreen/splitify.py:86]:
```suggestion
                raise commands.BadArgument('Squares must be greater than zero.')
```
I think this is more clear, and I don't like the idea of emojis in error messages.

File: bot/exts/evergreen/splitify.py
Comment (ToxicKidz) [inline bot/exts/evergreen/splitify.py:83]: This will be converted automatically, no need to convert it again.

File: bot/exts/evergreen/splitify.py
Comment (ToxicKidz) [inline bot/exts/evergreen/splitify.py:91]: I think this should be a constant.

**Review by ToxicKidz — CHANGES_REQUESTED** [review]
Thanks for the PR! Just couple suggestions.

File: bot/exts/evergreen/splitify.py
Comment (vincepy) [reply bot/exts/evergreen/splitify.py:80]: Alright, I'll do that.

File: bot/exts/evergreen/splitify.py
Comment (vincepy) [reply bot/exts/evergreen/splitify.py:83]: Alright, I'll do that.

File: bot/exts/evergreen/splitify.py
Comment (vincepy) [reply bot/exts/evergreen/splitify.py:86]: Well those errors get shown to users so I thought it would be more interactive and would look nicer, not so bald.

File: bot/exts/evergreen/splitify.py
Comment (vincepy) [reply bot/exts/evergreen/splitify.py:91]: Alright, I'll do that.

File: bot/exts/evergreen/splitify.py
Comment (vincepy) [reply bot/exts/evergreen/splitify.py:106]: I don't think so. Same thing in 8bitify was kept in the command.

Comment (Shivansh-007) [reply bot/exts/evergreen/splitify.py:105]: We are moving all image processing related commands into a single executor, #597 like you can see here.

I would suggest waiting until that PR gets merged and then moving this into it's group, so they will be inside the same executor. 

Or if the PR is not getting merged anytime soon, you can move it into a executor like toxic said.

cc: @ChrisLovering

File: bot/exts/evergreen/splitify.py
Comment (Shivansh-007) [inline bot/exts/evergreen/splitify.py:21]: Could you explain the algorithm you used for solving this task in the doc-string?

File: bot/exts/evergreen/splitify.py
Comment (Shivansh-007) [inline bot/exts/evergreen/splitify.py:50]: Could you explain the algorithm you used for solving this task in the doc-string?

File: bot/exts/evergreen/splitify.py
Comment (Shivansh-007) [inline bot/exts/evergreen/splitify.py:95]: You can directly use `ctx.author`, no need to add a wrapper around it.

File: bot/exts/evergreen/splitify.py
Comment (Shivansh-007) [inline bot/exts/evergreen/splitify.py:120]: This is against the style guide we follow, which can be seen [here](https://pythondiscord.com/pages/contributing/style-guide/)

To be more specific, this part:
> In the case where items contained in brackets need to be broken across multiple lines, items should be dropped to a new line after the opening bracket with an additional level of indentation. The closing bracket ends on it's own new line, on the same indentation level as the opening bracket.

```suggestion
            embed = discord.Embed(
              title=title,
              description=description
            )
```

File: bot/exts/evergreen/splitify.py
Comment (Shivansh-007) [inline bot/exts/evergreen/splitify.py:10]: Any specific reason for why you are limiting the squares to this particular number?

**Review by Shivansh-007 — CHANGES_REQUESTED** [review]
Good work on your first PR to sir-lancebot, if anything seems unclear to you, let me know, I would be happy to help you out 😄

File: bot/exts/evergreen/splitify.py
Comment (vincepy) [reply bot/exts/evergreen/splitify.py:105]: I will try an executor.

File: bot/exts/evergreen/splitify.py
Comment (vincepy) [reply bot/exts/evergreen/splitify.py:21]: Is there an issue with it or are you just personally interested?

File: bot/exts/evergreen/splitify.py
Comment (vincepy) [reply bot/exts/evergreen/splitify.py:50]: Is there an issue with it or are you just personally interested?

File: bot/exts/evergreen/splitify.py
Comment (vincepy) [reply bot/exts/evergreen/splitify.py:95]: Alright, I'll do that.

File: bot/exts/evergreen/splitify.py
Comment (vincepy) [reply bot/exts/evergreen/splitify.py:120]: Alright, I'll do that.

File: bot/exts/evergreen/splitify.py
Comment (Shivansh-007) [reply bot/exts/evergreen/splitify.py:21]: It would help one going through the code and understanding how it is done.

Say, one contributor wants to refactor this code and change the implementation a bit, he would need to understand this code for doing that, when you put the doc-strings in there, it would help them do it.

File: bot/exts/evergreen/splitify.py
Comment (vincepy) [reply bot/exts/evergreen/splitify.py:10]: A greater number than 193,600 causes me PIL errors, I did a lot of testing. If I wanted, I could probably recode it to work with bigger numbers, but since 193,000 squares (split pieces) is already so many for a single Discord PFP, I decided it would serve as a great cap.

Comment (Shivansh-007) [reply bot/exts/evergreen/splitify.py:10]: Ah ok, thanks!

File: bot/exts/evergreen/splitify.py
Comment (vincepy) [reply bot/exts/evergreen/splitify.py:21]: Well may I explain myself then.

1. I get the width and the height of the Image passed to the function.

2. I get the root of a number of squares passed which I call xy. Reason: if let's say 25 squares were passed, that is the total squares (split pieces) that the image is supposed to be split into. As you know, a 2D shape has a height and a width, and for this case I think of it as rows and columns. Rows multiplied by columns is equal to the passed squares. To get rows and columns, since in this case each square is identical, rows are equal to columns and I treat the Image as a square-shaped, I get the root out of the squares passed.

3. Now width and height are both of the original Image, Discord PFP, so when it comes to forming the squares (split pieces), I divide the original image's height and width by the xy. In a case of 25 squares passed, xy would be 5, so if an image was 250x300, x_frac would be 50 and y_frac - 60. Note: x_frac stands for a fracture of width. Reason I called it so is because it is shorter to use x for width in mind and then I just used half of the word fracture, same applies to y_frac, just height instead of width. x_frac and y_frac are width and height of a single square (split piece).

4. With `left, top, right, bottom, = 0, 0, x_frac, y_frac`, I set these variables to create the initial square (split piece). Explanation: all of these 4 variables start at the top left corner of the Image, by adding value to right and bottom, I am creating the initial square (split piece).

5. In the for loop, I keep adding those squares (split pieces) in row and once `(index + 1) % xy == 0` is True, I add to top and bottom to lower them and reset right and left to recreate the initial space between them, forming a square (split piece), I also add the newly created square (split piece) into the new_imgs list where I store them. I keep repeating this process till all 25 squares get added to the list.

Hopefully this is comprehendible.

Comment (Shivansh-007) [reply bot/exts/evergreen/splitify.py:21]: I did understand the code originally but Thanks for the awesome explanation, what I was asking for to include this explanation in your code as comments and doc-strings so the algorithm can be much clear to other contributors.

Comment (vincepy) [reply bot/exts/evergreen/splitify.py:21]: Alright, I'll do that.

File: bot/exts/evergreen/splitify.py
Comment (vincepy) [reply bot/exts/evergreen/splitify.py:50]: Here it is.

1. I shuffle passed imgs. Reason: to randomize the squares (split pieces).

2. I get a single square (split piece) out of the list and define single_width as the square's width and single_height as the square's height.

3. I get the root of type integer of the number of imgs (split pieces) in the list and call it multiplier. I then proceed to calculate total height and width of the new image that I am creating using the same multiplier.

4. I then define new_img as the image that I am creating, using the previously obtained total_width and total_height.

5. Now I define width_multiplier as well as height with values of 0. These will be used to correctly position squares (split pieces) onto the new_img canvas.

6. Similar to how in the split_image function, I get the root of number of imgs (split pieces) in the list. In split_image function, it was the passed squares (number of squares) instad of a number of imgs in the list that I got the square of here.

7. In the for loop, as I iterate, I multiply single_width by width_multiplier to correctly position a square (split piece) width wise. I then proceed to paste the newly positioned square (split piece) onto the new_img. I increase the width_multiplier by 1 every iteration so the image wouldn't get pasted in the same spot and the positioning would move accordingly. I make sure to increase the width_multiplier before the check, which checks if the end of a row has been reached, - (index + 1) % pieces == 0, so after it, if it was True, width_multiplier would have been reset to 0 (start of the row). If the check returns True, the height gets increased by a single square's (split piece) height to lower the positiong height wise and, as I just mentioned, the width_multiplier gets reset to 0 and width will then be calculated from the start of the new row. The for loop finishes once all the squres were positioned accordingly.

8. Finally, I return the new_img, the randomized squares (split pieces) stitched back into the format of the original image - user's PFP.

**Review by Kronifer — APPROVED** [review]
Spectacular first pr for sir lancebot. Great work!

**Comment by vincepy** [comment]
> Just a small consistency change here, below you use a space between `193,000` and :nerd: which should be replicated here for a consistent look
> 
> Thanks github for not showing where :/
> around line 86

Sorry, I don't understand what you meant. Could you try showing in code?

**Comment by Shivansh-007** [comment]
Hey @vincepy! Could you resolve the changes which you have implemented, this would help reviewers know which changes have been implemented which have not.

Thanks!

File: bot/exts/evergreen/splitify.py
Comment (ToxicKidz) [inline bot/exts/evergreen/splitify.py:24]:
```suggestion
    def split_image(img: Image.Image, squares: int) -> list:
```
Currently you're type hinting to a module.

File: bot/exts/evergreen/splitify.py
Comment (ToxicKidz) [inline bot/exts/evergreen/splitify.py:84]:
```suggestion
    def join_images(imgs: list) -> Image.Image:
```
Same thing here.

**Review by ToxicKidz — CHANGES_REQUESTED** [review]

**Comment by vincepy** [comment]
> Hey @vincepy! Could you resolve the changes which you have implemented, this would help reviewers know which changes have been implemented which have not.
> 
> Thanks!

Done.

File: bot/exts/evergreen/splitify.py
Comment (vincepy) [reply bot/exts/evergreen/splitify.py:181]: Could you explain what do you mean here?

File: bot/exts/evergreen/splitify.py
Comment (vincepy) [reply bot/exts/evergreen/splitify.py:84]: Done.

File: bot/exts/evergreen/splitify.py
Comment (vincepy) [reply bot/exts/evergreen/splitify.py:24]: Done.

File: bot/exts/evergreen/splitify.py
Comment (ToxicKidz) [reply bot/exts/evergreen/splitify.py:181]: The bot instance has a loop attribute.

Comment (vincepy) [reply bot/exts/evergreen/splitify.py:181]: Alright, done.

**Review by vcokltfre — APPROVED** [review]
Looks great! :shipit:

**Review by ToxicKidz — APPROVED** [review]
It's a great PR, thanks! :shipit:

**Review by Shivansh-007 — APPROVED** [review]

**Comment by Shivansh-007** [comment]
Whoops, missed out one change, everywhere you use `I` in the code, first-person (and second-person) pronouns should be avoided, especially in doc strings. You can use `it` rather,

**Comment by vincepy** [comment]
> Whoops, missed out one change, everywhere you use `I` in the code, first-person (and second-person) pronouns should be avoided, especially in doc strings. You can use `it` rather,

Done.

File: bot/exts/evergreen/splitify.py
Comment (jb3) [inline bot/exts/evergreen/splitify.py:17]: Splitify is a vague term.
```suggestion
    """Turn your avatar into a mosaic!"""
```

File: bot/exts/evergreen/splitify.py
Comment (jb3) [inline bot/exts/evergreen/splitify.py:25]:
```suggestion
        Split an image into a selection of squares, specified by the squares argument.
```

File: bot/exts/evergreen/splitify.py
Comment (jb3) [inline bot/exts/evergreen/splitify.py:89]:
```suggestion
        1. Shuffles the passed imgs list to randomise the pieces.
```

File: bot/exts/evergreen/splitify.py
Comment (jb3) [inline bot/exts/evergreen/splitify.py:83]: Why not just call this `images`?
```suggestion
    def join_images(images: list) -> Image.Image:
```

File: bot/exts/evergreen/splitify.py
Comment (jb3) [inline bot/exts/evergreen/splitify.py:168]: I think we should use a different name for this.
```suggestion
    @commands.command(name='mosaic')
    async def mosaic_command(self, ctx: commands.Context, squares: int = 16) -> None:
```

File: bot/exts/evergreen/splitify.py
Comment (jb3) [inline bot/exts/evergreen/splitify.py:172]: As mentioned, the emoji here just looks weird. We should use no exclamation at all or a ascii excalamation point, we can also drop the code formatting.
```suggestion
                raise commands.BadArgument('Squares must be a positive number')
```

File: bot/exts/evergreen/splitify.py
Comment (jb3) [inline bot/exts/evergreen/splitify.py:175]:
```suggestion
                raise commands.BadArgument('Squares must be a perfect square')
```

File: bot/exts/evergreen/splitify.py
Comment (jb3) [inline bot/exts/evergreen/splitify.py:178]:
```suggestion
                raise commands.BadArgument(f'Number of squares cannot be higher than {MAX_SQUARES:,}.')
```

File: bot/exts/evergreen/splitify.py
Comment (jb3) [inline bot/exts/evergreen/splitify.py:193]: This is a bit weird gramatically.
```suggestion
                description = 'Here is your avatar. I think it looks a bit *puzzling*'
```

**Review by jb3 — CHANGES_REQUESTED** [review]
This is a pretty neat command. I've got some feedback added as code comments as well as the items below:

## Command naming

I think the name `splitify` is a bit weird and hard to understand what it does. I'd recommend either switching all references to either `puzzle`, `mosaic` or something along those lines.

## Maximum squares

Here are two images generated by the command:

 193,000 squares | 10,000 squares
:---------------:|:----------------:
image|image

10,000 produces a more interesting image compared to 193,000 at a much lower resource utilisation and generation time. I think we should use a lower limit between 10,000 and 40,000.

**Pull request closed without merging**
